Rule tree, v1

emilio opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] There are tests for these changes OR

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

This patch introduces infrastructure for the rule tree, and constructs it.

We don't use it yet, nor have good heuristics for GC'ing it, but this should not
happen anymore once we store the rule node reference in the node.

I haven't messed up with memory orders because I want to do a try run with it,
then mess with them.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13202)

<!-- Reviewable:end -->

emilio committed:
Concurrent rule tree, v1

This patch introduces infrastructure for the rule tree, and constructs it.

We don't use it yet, nor have good heuristics for GC'ing it, but this should not
happen anymore once we store the rule node reference in the node.

I haven't messed up with memory orders because I want to do a try run with it,
then mess with them.

Take down the ApplicableDeclarationsCache, use the rule tree for doing the cascade.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/lib.rs, components/style/parallel.rs, components/style/values/specified/mod.rs, components/style/selector_matching.rs, components/style/properties/properties.mako.rs, components/style/data.rs, components/style/thread_state.rs, components/style/rule_tree/mod.rs, components/style/animation.rs, components/style/matching.rs
 * @kichjang: components/script_layout_interface/wrapper_traits.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit f1f6891 with merge dfd58bf...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2910)

SimonSapin commented on components/style/animation.rs:
```diff
@@ -384,12 +384,9 @@ fn compute_style_for_animation_step(context: &SharedStyleContext,
-                mixed_declarations: declarations.clone(),
-                importance: Importance::Normal,
-                source_order: 0,
-                specificity: ::std::u32::MAX,
```
Isn’t max specificity here important?

emilio replied:
I don't think so, we only pass it to `cascade`, that doesn't use it at all.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13205) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit d333633 with merge e3c3d7d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3099)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/setpropertypriority.html
      └   → /_mozilla/css/setpropertypriority.html 059350c3fab2914549d382466f600fa6be9e9599
    /_mozilla/css/setpropertypriority_ref.html f565a85a358e1e10906e0a87eb234acff04e9bfb
    Testing 059350c3fab2914549d382466f600fa6be9e9599 == f565a85a358e1e10906e0a87eb234acff04e9bfb

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit f2251d4 with merge 8992c37...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3054)

highfive commented:
▶ ERROR [expected OK] /webgl/conformance-1.0.3/conformance/ogles/GL/build/build_009_to_016.html
      └   → gl.getProgramInfoLog is not a function

jdm commented:
@bors-servo: retry
* #13165

bors-servo commented:
:hourglass: Trying commit f2251d4 with merge 9d01a6a...

bors-servo commented:
:broken_heart: Test failed - [arm64](http://build.servo.org/builders/arm64/builds/2885)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13213) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit b23cbbc with merge f7b2397...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3086)

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit b23cbbc with merge 0da9621...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2931), [arm64](http://build.servo.org/builders/arm64/builds/2947), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3000), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3169), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3006), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3000), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3118), [windows-dev](http://build.servo.org/builders/windows-dev/builds/894)

emilio commented:
r? @SimonSapin

pcwalton commented on components/style/rule_tree/mod.rs:
```diff
@@ -0,0 +1,491 @@
+
+    /// The next item in the rule tree free list, that starts on the root node.
+    next_free: AtomicPtr<RuleNode>,
+}
```
Can we fuzz this, or factor this out into a generic, or something like that? The amount and trickiness of the unsafe code here concerns me a little bit. (I'm not going to stand in the way of this work obviously, but would just like to know if there's anything we can do to harden this.)

emilio replied:
Yeah, I thought about making it generic, but that might get tricky. I think I can write a few unit tests that stress this a bit, or something.

And yeah, totally agree in the documentation part, this is ready for review, but needs docs before landing.

pcwalton commented on components/style/rule_tree/mod.rs:
```diff
@@ -0,0 +1,491 @@
+            node.refcount.fetch_sub(1, Ordering::SeqCst) == 1
+        };
+
+        if should_drop {
```
nit: avoid rightward drift with:

    if !should_drop {
        return
    }

pcwalton commented on components/style/rule_tree/mod.rs:
```diff
@@ -0,0 +1,491 @@
+}
+
+struct RuleChildrenListIter {
+    current: Option<WeakRuleNode>,
```
Nobody can delete children while this is being iterated over, right?

emilio replied:
Right, the children are only deleted in the layout thread.

pcwalton commented on components/style/rule_tree/mod.rs:
```diff
@@ -0,0 +1,491 @@
+                    // Now we know we're in the correct position in the child list,
+                    // we can set the back pointer, knowing that this will only be
+                    // accessed again in a single-threaded manner when we're
+                    // sweeping possibly dead nodes.
```
Can you add a comment to `prev` specifying this very special memory constraint? This is *tricky*.

emilio replied:
Sure, in general, I need to write a lot more docs before landing this.

pcwalton commented on components/style/rule_tree/mod.rs:
```diff
@@ -0,0 +1,491 @@
+    ///
+    /// This is expected to be called before freeing the node from the free
+    /// list.
+    unsafe fn remove_from_child_list(&self) {
```
In debug mode, could you set a `sweeping_in_progress` flag here and have all of the list methods check that flag and panic if it's set?

I know it's paranoid, but if it ever trips in the test suite we'll be *really* glad that we added that flag.

emilio replied:
Sure, though that'd require #13127

notriddle replied:
It's landed.

emilio replied:
Thanks for the reminder. I need to do a huge rebase on this, I'll add this when I have the time.

Manishearth commented:
Reviewed 3 of 19 files at r1.
Review status: 1 of 24 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/style/rule_tree/mod.rs, line 53 at r1](https://reviewable.io:443/reviews/servo/servo/13202#-KS_m-0tTHxuLiKucqmA:-KS_m-0uttg4sCTmZgVB:byncw4k) ([raw file](https://github.com/servo/servo/blob/b0260c640803b246ba10e0ea53abe90e17b2ecb9/components/style/rule_tree/mod.rs#L53)):*
> ```Rust
>         where I: Iterator<Item=(&'a Arc<StyleRule>, Importance)>
>     {
>         let root = self.root.clone();
> ```

I don't think this clone is necessary

---

*[components/style/rule_tree/mod.rs, line 245 at r1](https://reviewable.io:443/reviews/servo/servo/13202#-KS_vKiRbrLuY5XwWr3u:-KS_vKiSgJP8poH3soNI:b-te96ef) ([raw file](https://github.com/servo/servo/blob/b0260c640803b246ba10e0ea53abe90e17b2ecb9/components/style/rule_tree/mod.rs#L245)):*
> ```Rust
>         let ptr = Box::into_raw(n);
> 
>         debug!("Creating rule node: 0x{:x}", ptr as usize);
> ```

`{:p}`

---

*[components/style/rule_tree/mod.rs, line 281 at r1](https://reviewable.io:443/reviews/servo/servo/13202#-KS_zjWFGeHQpWf0HMN-:-KS_zjWGVv27MzjGWlkQ:b-igqyy9) ([raw file](https://github.com/servo/servo/blob/b0260c640803b246ba10e0ea53abe90e17b2ecb9/components/style/rule_tree/mod.rs#L281)):*
> ```Rust
>         let mut last = None;
>         let mut iter = self.get().children.iter();
>         while let Some(current) = iter.next() {
> ```

Any reason why this isn't a for loop?

---

*[components/style/rule_tree/mod.rs, line 424 at r1](https://reviewable.io:443/reviews/servo/servo/13202#-KSfG9VhbOv6ldlc242H:-KSfG9Vi1WlgaOV1HlOA:b-bhmjon) ([raw file](https://github.com/servo/servo/blob/b0260c640803b246ba10e0ea53abe90e17b2ecb9/components/style/rule_tree/mod.rs#L424)):*
> ```Rust
> 
>     fn upgrade(&self) -> StrongRuleNode {
>         debug!("Upgrading weak node: 0x{:?}", self.ptr());
> ```

`{:p}` again

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
rule tree safety stuff lgtm, sorry I didn't get to this before.

would like more eyes on this ofc.

---

Reviewed 2 of 19 files at r1, 1 of 18 files at r2.
Review status: 3 of 24 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202#-:-KTKXp3AMPw_sgRtkHDH:b-108ne1)*
<!-- Sent from Reviewable.io -->

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 7b7a3de with merge 16654b2...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3621)

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 8e894f9 with merge 87912ba...

bors-servo commented:
:boom: Test timed out

emilio commented:
@bors-servo: try- retry

bors-servo commented:
:hourglass: Trying commit 8e894f9 with merge 8056122...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3548), [arm64](http://build.servo.org/builders/arm64/builds/3559), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3625), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/256), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/257), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3626), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3604), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/306), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/308), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1515)

emilio commented:
@bholley: @manishearth: @SimonSapin: Assuming we still want/need a rule tree, which I think is true, how much would it cost to get it reviewed and landed? It was a pain to rebase :-)

bholley commented:
We definitely still want/need the rule tree! Sounds like @Manishearth has signed off on the safety, and we're just waiting for @SimonSapin's review?

Manishearth commented:
Yes, I'm fine with the safety (and the style bits look good too), but I want someone else to look at the style bits.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13839) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13863) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2a2305c with merge e7f793b...

bors-servo commented:
:broken_heart: Test failed - [windows-dev](http://build.servo.org/builders/windows-dev/builds/1677)

KiChjang commented:
```
error[E0308]: mismatched types
  --> C:\buildbot\slave\windows-dev\build\tests\unit\style\selector_matching.rs:26:20
   |
26 |             block: PropertyDeclarationBlock {
   |                    ^ expected struct `std::sync::Arc`, found struct `style::properties::PropertyDeclarationBlock`
   |
   = note: expected type `std::sync::Arc<parking_lot::RwLock<style::properties::PropertyDeclarationBlock>>`
   = note:    found type `style::properties::PropertyDeclarationBlock`

error[E0308]: mismatched types
  --> C:\buildbot\slave\windows-dev\build\tests\unit\style\selector_matching.rs:39:29
   |
39 |                 style_rule: rule.clone(),
   |                             ^^^^^^^^^^^^ expected struct `parking_lot::RwLock`, found struct `style::stylesheets::StyleRule`
   |
   = note: expected type `std::sync::Arc<parking_lot::RwLock<style::stylesheets::StyleRule>>`
   = note:    found type `std::sync::Arc<style::stylesheets::StyleRule>`

error: aborting due to 2 previous errors

error: Could not compile `style_tests`.
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13913) made this pull request unmergeable. Please resolve the merge conflicts.

bholley commented on components/style/data.rs:
```diff
@@ -39,14 +40,18 @@ pub struct NodeStyles {
     pub primary: Arc<ComputedValues>,
 
+    /// The rule node representing the last rule matched for this node.
+    pub rule_node: StrongRuleNode,
```
Can you group these two into:

pub struct StyleResult {
    pub rule_node: Arc<ServoRuleNode>,
    pub computed_values: Arc<ServoComputedValues>,
}

That will match the algorithm at [1], and allow us to make PseudoStylesInner a map between PseudoElement and StyleResult. This will also allow us to transfer ownership of both to the frame.

[1] https://docs.google.com/document/d/1TcTAQMm-jIiNgzkDgpO04Pja_aQNRNkrD0CNCCYfbo8

bholley commented:
Is there anything blocking this from landing? I'm getting to the point where this would be helpful to make progress in the incremental restyle algorithm. :-)

bholley commented:
You can probably just gate it on cfg_attr(feature = "servo", ....) like we do for HeapSizeOf.

SimonSapin committed:
rule tree: Remove unsound transmute()

RuleTreeDeclarationsIterator would yield &PropertyDeclaration
with the lifetime of the rule tree, but the reference would only
be valid as long as the iterator was holding the corresponding lock.

The lock would be released when the iterator’s `.next()` method
moves to the next rule tree node, or when  the iterator is dropped.

Also change apply_declaration to not require a `Clone` iterator
(since closures unfortunately don’t implement `Clone`).
Instead have it take a callable that returns a fresh iterator.

SimonSapin commented:
`transmute` in `RuleTreeDeclarationsIterator` is unsound. It extends the lifetime of `&PropertyDeclaration` block to that of the rule tree, but that reference is only valid as long as the lock is held.

I’ve pushed a commit at https://github.com/emilio/servo/compare/rule-tree...servo:rule-tree that removes this iterator and makes things work differently. In that commit, `cascade()` allocates a `Vec` that contains lock guards. This is necessary because `apply_declarations()`, for cascading custom properties, uses borrows that requires these locks to be held the whole time. This also avoids unlocking/re-locking repeatedly. (`apply_declarations()` iterates through declarations three times.)

SimonSapin commented:
Reviewed 1 of 19 files at r1, 1 of 18 files at r2, 26 of 28 files at r3.
Review status: 23 of 25 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[components/style/declarations_iterators.rs, line 14 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KV-QU1MMNpjE_MkGeUp:-KV-QU1MMNpjE_MkGeUq:busu49i) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/declarations_iterators.rs#L14)):*
> ```Rust
> #[derive(Clone)]
> pub struct SimpleDeclarationsIterator<'a> {
>     declarations: &'a [(PropertyDeclaration, Importance)],
> ```

(This could have been simpler with `std::slice::Iter<'a, …>`, but I’m removing it in a next commit.)

---

*[components/style/matching.rs, line 56 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjum6w4pZgrOdWCgCF:-KVjum6w4pZgrOdWCgCG:bb76mu4) ([raw file](https://github.com/servo/servo/blob/c8321edc27ed7125181262e0341e66825951f634/components/style/matching.rs#L56)):*
> ```Rust
> 
> pub struct ApplicableDeclarations {
>     pub normal: SmallVec<[ApplicableDeclarationBlock; 16]>,
> ```

Why?

---

*[components/style/selector_matching.rs, line 311 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjwIrI_UgeMhOOwnsy:-KVjwIrJ6CyFzQpXyOF9:bblhll7) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/selector_matching.rs#L311)):*
> ```Rust
> 
>         let rule_node =
>             self.rule_tree.insert_ordered_rules(declarations.into_iter().map(|a| (a.source.clone(), a.importance)));
> ```

Is `.clone()` (and even `map`) here necessary? Since `.into_iter()` gives ownership of items.

---

*[components/style/selector_matching.rs, line 899 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjwlU9Eqd4iwppeX_F:-KVjwlU9Eqd4iwppeX_G:bqlffzb) ([raw file](https://github.com/servo/servo/blob/c8321edc27ed7125181262e0341e66825951f634/components/style/selector_matching.rs#L899)):*
> ```Rust
> #[derive(Debug, Clone)]
> pub struct ApplicableDeclarationBlock {
>     /// Contains declarations of either importance, but only those of self.importance are relevant.
> ```

I think this comment should be kept.

---

*[components/style/properties/properties.mako.rs, line 1643 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjyPdzLDdKdowN2Pgr:-KVjyPdzLDdKdowN2Pgs:bcmmgya) ([raw file](https://github.com/servo/servo/blob/c8321edc27ed7125181262e0341e66825951f634/components/style/properties/properties.mako.rs#L1643)):*
> ```Rust
>                         PropertyDeclaration::Float(_) |
>                         PropertyDeclaration::TextDecoration${'' if product == 'servo' else 'Line'}(_) |
>                         PropertyDeclaration::WritingMode(_)
> ```

Looks like `WritingMode` is missing in the new code. (Probably lost in rebase?)

---

*[components/style/properties/properties.mako.rs, line 1655 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjyXGGXxnhL5pt1AkB:-KVjyXGGXxnhL5pt1AkC:b-hse17f) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/properties/properties.mako.rs#L1655)):*
> ```Rust
> /// NOTE: This function expects the declaration with more priority to appear
> /// first.
> pub fn apply_declarations<'a, I>(viewport_size: Size2D<Au>,
> ```

Please move this next to `cascade()`, it’ll make the diff easier to read.

---

*[components/style/properties/properties.mako.rs, line 1711 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVjxjms5aLGP6q65NY1:-KVjxjmtA96V_nGjHqKE:bpvkiob) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/properties/properties.mako.rs#L1711)):*
> ```Rust
> 
>     // Declaration blocks are stored in increasing precedence order, we want
>     // them in decreasing order here.
> ```

This comment is not relevant anymore.

---

*[components/style/rule_tree/mod.rs, line 1 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KV-Xkx4g6SZiVS_KbGR:-KV-Xkx5zt3TmSn10mKJ:bnme9hl) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/rule_tree/mod.rs#L1)):*
> ```Rust
> /* This Source Code Form is subject to the terms of the Mozilla Public
> ```

Does this file need to be `rule_tree/mod.rs` rather than `rule_tree.rs`?

---

*[components/style/values/specified/mod.rs, line 218 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KV-Jy6g9ltWlTP7QmuN:-KV-Jy6hBDTtL7FkU6L9:bbul96o) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/values/specified/mod.rs#L218)):*
> ```Rust
>             Length::ViewportPercentage(length) => length.to_css(dest),
>             Length::Calc(ref calc, _) => calc.to_css(dest),
>             /* This should only be reached from style dumping code */
> ```

I don’t like relying on this "should", but the alternative probably requires implementing `Debug` on everything which seems busy work not worth it.

If this gonna be part of a `ToCss` impl though, at least make it look like CSS. Something like `"{}-servo-character-width"` for the formatting string. (A vendor-prefixed unit.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202)*
<!-- Sent from Reviewable.io -->

SimonSapin committed:
Simplify rule_tree::StyleSourceGuardHandle

SimonSapin committed:
rule tree: more descriptive field names

SimonSapin commented:
Reviewed 2 of 28 files at r3.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[components/style/rule_tree/mod.rs, line 499 at r3](https://reviewable.io:443/reviews/servo/servo/13202#-KVofoDQUvaaZc2-_yHZ:-KVofoDQUvaaZc2-_yH_:bz7mt7l) ([raw file](https://github.com/servo/servo/blob/2a2305c4ea6ba64f5d0a6cf55e85128b135639ac/components/style/rule_tree/mod.rs#L499)):*
> ```Rust
>             // plus `decl` has a stable address (because it's in an Arc), so we
>             // need to transmute here.
>             let decl: &'a PropertyDeclaration = unsafe { ::std::mem::transmute(decl) };
> ```

As mentioned in a PR comment, this `transmute` is unsound. https://github.com/emilio/servo/pull/1 moves things around to make it unnecessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Carrying review from Emilio in https://github.com/emilio/servo/pull/1

---

Reviewed 6 of 6 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202#-:-KVoxbuEqj_4nEesqXuP:b-dtlqwx)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
@bors-servo r=SimonSapin,emilio

---

Reviewed 13 of 22 files at r8, 4 of 6 files at r9, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13202#-:-KVpNnBqa9JL5Hv_ubAr:bs2evme)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 1a18161 has been approved by `SimonSapin,emilio`

<!-- @bors-servo r=SimonSapin,emilio 1a18161006c67424750943e1b29d3f3aead92d5e -->

SimonSapin commented:
@bors-servo p=10

Rebase was tricky.

emilio commented:
@bors-servo try

Thanks simon for the work and the reviews here! I think he was ok with landing it in its current state?

bors-servo commented:
:hourglass: Trying commit 1a18161 with merge 55ae5e1...

emilio commented:
My browser wasn't updating the page apparently, sigh.

@bors-servo r=SimonSapin,Manishearth,emilio p=10

^ just in case

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request is currently being tested. If there's no response from the continuous integration service, you may use `retry` to trigger a build again.

bors-servo commented:
:pushpin: Commit 1a18161 has been approved by `SimonSapin,Manishearth,emilio`

<!-- @bors-servo r=SimonSapin,Manishearth,emilio 1a18161006c67424750943e1b29d3f3aead92d5e -->

notriddle commented:
http://build.servo.org/homu/queue/servo says it's pending. GitHub says it's trying.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/675)

highfive commented:
▶ TIMEOUT [expected CRASH] /_mozilla/mozilla/sigsegv.html
      │ 
      │ VMware, Inc.
      │ Gallium 0.4 on softpipe
      └ 3.3 (Core Profile) Mesa 12.0.1

SimonSapin commented:
@bors-servo retry

* #14067

SimonSapin commented:
@bors-servo force

Trying to get the queue unstuck, http://build.servo.org/waterfall is showing no activity.

SimonSapin commented:
@bors-servo retry force

Trying to get the queue unstuck, http://build.servo.org/waterfall is showing no activity.

bors-servo commented:
:hourglass: Testing commit 1a18161 with merge 151fb7e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/677)

highfive commented:
▶ TIMEOUT [expected CRASH] /_mozilla/mozilla/sigsegv.html
      │ 
      │ VMware, Inc.
      │ Gallium 0.4 on softpipe
      └ 3.3 (Core Profile) Mesa 12.0.1

emilio commented:
@bors-servo retry

    #14067

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3903), [arm64](http://build.servo.org/builders/arm64/builds/3911), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4000), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/640), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/644), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3987), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3982), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/667), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1910) are reusable. Rebuilding only [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/677)...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:hourglass: Testing commit 1a18161 with merge ac04759...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3906), [arm64](http://build.servo.org/builders/arm64/builds/3914), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4003), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/643), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/647), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3990), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3985), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/670), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/681), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1913)
